private/bufpkg/bufmodule/bufmodulecache/cas_module_cacher.go

```diff
@@ -0,0 +1,272 @@
+			return fmt.Errorf("manifest digest mismatch: expected=%q, found=%q", modulePinDigest.String(), digest.String())
+		}
+	}
+	moduleBasedir := normalpath.Join(modulePin.Remote(), modulePin.Owner(), modulePin.Repository())
+	// Write blobs
+	writtenDigests := make(map[string]struct{})
+	for _, path := range moduleManifest.Paths() {
+		blobDigest, found := moduleManifest.DigestFor(path)
+		if !found {
+			return fmt.Errorf("failed to find digest for path=%q", path)
+		}
+		blobDigestStr := blobDigest.String()
+		if _, ok := writtenDigests[blobDigestStr]; ok {
+			continue
+		}
+		blob, found := module.BlobSet().BlobFor(blobDigestStr)
+		if !found {
+			return fmt.Errorf("blob not found for path=%q, digest=%q", path, blobDigestStr)
+		}
+		if err := c.writeBlob(ctx, moduleBasedir, blob); err != nil {
+			return err
+		}
+		writtenDigests[blobDigestStr] = struct{}{}
+	}
+	// Write manifest
+	if err := c.writeBlob(ctx, moduleBasedir, manifestBlob); err != nil {
+		return err
+	}
+	// Write commit
+	commitPath := normalpath.Join(moduleBasedir, commitsDir, modulePin.Commit())
+	if err := c.atomicWrite(ctx, strings.NewReader(manifestBlob.Digest().String()), commitPath); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *casModuleCacher) readBlob(
+	ctx context.Context,
+	moduleBasedir string,
+	digest *manifest.Digest,
+) (_ manifest.Blob, retErr error) {
+	hexDigest := digest.Hex()
+	blobPath := normalpath.Join(moduleBasedir, blobsDir, hexDigest[:2], hexDigest[2:])
+	contents, err := c.loadPath(ctx, blobPath)
+	if err != nil {
+		return nil, err
+	}
+	blob, err := manifest.NewMemoryBlob(*digest, contents, manifest.MemoryBlobWithDigestValidation())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create blob from path %s: %w", blobPath, err)
+	}
+	return blob, nil
+}
+
+func (c *casModuleCacher) validateBlob(
+	ctx context.Context,
+	moduleBasedir string,
+	digest *manifest.Digest,
+) (bool, error) {
+	hexDigest := digest.Hex()
+	blobPath := normalpath.Join(moduleBasedir, blobsDir, hexDigest[:2], hexDigest[2:])
+	f, err := c.bucket.Get(ctx, blobPath)
+	if err != nil {
+		return false, err
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			c.logger.Debug("err closing blob", zap.Error(err))
+		}
+	}()
+	digester, err := manifest.NewDigester(digest.Type())
+	if err != nil {
+		return false, err
+	}
+	cacheDigest, err := digester.Digest(f)
+	if err != nil {
+		return false, err
+	}
+	return digest.String() == cacheDigest.String(), nil
+}
+
+func (c *casModuleCacher) readManifest(
+	ctx context.Context,
+	moduleBasedir string,
+	manifestDigest *manifest.Digest,
+) (_ *manifest.Manifest, retErr error) {
+	blob, err := c.readBlob(ctx, moduleBasedir, manifestDigest)
+	if err != nil {
+		return nil, err
+	}
+	f, err := blob.Open(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		retErr = multierr.Append(retErr, f.Close())
+	}()
+	moduleManifest, err := manifest.NewFromReader(f)
```

Comment:
Feels weird to be reading the blob twice here. It's an interface issue I guess. Food for thought.

Reply:
We ended up doing it this way because we first validate the blob's contents against the digest before attempting to consume it as a manifest. I don't think we are really reading it twice here since the second open just returns an in-memory view of the blob's contents.